public/PROPOSAL/math/Vector3D.hpp

```diff
@@ -0,0 +1,216 @@
+template <typename T> class Spherical3D;
+template <typename T> class Cartesian3D : public Vector3D<T> {
+public:
+    Cartesian3D(array<T, 3> val)
+        : Vector3D<T>(val){};
+
+    Cartesian3D operator+=(const Cartesian3D&);
+
+    template <typename U>
+    friend Cartesian3D<U> operator*(Cartesian3D<U> vec, double factor);
+    template <typename U>
+    friend U operator*(const Cartesian3D<U>&, Cartesian3D<U>&);
+    template <typename U>
+    friend Cartesian3D<U> operator+(Cartesian3D<U>, const Cartesian3D<U>&);
+    template <typename U>
+    friend Cartesian3D<U> operator-(Cartesian3D<U>, const Cartesian3D<U>&);
+    template <typename U>
+    friend Cartesian3D<U> cross_prod(
+        const Cartesian3D<U>&, const Cartesian3D<U>&);
+
+    T magnitude() const override;
+    Spherical3D<T> ToSpherical() const;
+};
+} // namespace PROPOSAL
+
+namespace PROPOSAL {
+template <typename T> class Spherical3D : public Vector3D<T> {
+public:
+    Spherical3D(array<T, 3> val)
+        : Vector3D<T>(val){};
+
+    template <typename U>
+    friend Spherical3D<U> operator*(Spherical3D<U> vec, double factor);
+    template <typename U>
+    friend U operator*(const Spherical3D<U>&, const Spherical3D<U>&);
+
+    T magnitude() const override;
+    Cartesian3D<T> ToCartesian() const;
+};
+} // namespace PROPOSAL
+
+namespace PROPOSAL {
+template <typename T> bool Vector3D<T>::operator==(const Vector3D&& rhs) const
+{
+    for (auto i : indices(values)) {
+        if (values[i] != rhs[i])
+            return false;
+    }
+    return true;
+}
+
+template <typename T> T& Vector3D<T>::operator[](size_t idx)
+{
+    return values[idx];
+}
+
+template <typename U>
+std::ostream& operator<<(std::ostream& os, const Vector3D<U>& vec)
+{
+    os << "================ Vector3D (" << &vec << ") ================\n";
+    os << "x0: " << vec.GetArray()[0] << "\t";
+    os << "x1: " << vec.GetArray()[1] << "\t";
+    os << "x2: " << vec.GetArray()[2] << "\n";
+    os << "===========================================================\n";
+    return os;
+}
+
+template <typename T>
+Cartesian3D<T> operator*(Cartesian3D<T> vec, double factor)
+{
+    for (auto& i : Vector3D<T>::values)
+        i *= factor;
+    return vec;
+}
+
+template <typename U>
+Cartesian3D<U> operator*(
+    const Cartesian3D<U>& l_vec, const Cartesian3D<U>& r_vec)
+{
+    U accum = 0;
+    for (auto i : indices(l_vec.values))
+        accum += l_vec[i] * r_vec[i];
+    return accum;
+}
+
+template <typename U>
+Cartesian3D<U> operator+(Cartesian3D<U> l_vec, Cartesian3D<U>& r_vec)
+{
+    for (size_t i = 0; i < 3; ++i)
+        l_vec[i] += r_vec[i];
+    return l_vec;
+}
+
+template <typename U>
+Cartesian3D<U> operator-(Cartesian3D<U> l_vec, const Cartesian3D<U>& r_vec)
+{
+    for (size_t i = 0; i < 3; ++i)
+        l_vec[i] -= r_vec[i];
+    return l_vec;
+}
+
+template <typename T>
+Cartesian3D<T> Cartesian3D<T>::operator+=(const Cartesian3D<T>& rhs)
+{
+    return *this + rhs;
```

Comment:
same error as last time! 

operator+= should mutate in place and return a reference. This creates unecessary copies.